stream_alert/classifier/parsers.py

```diff
@@ -0,0 +1,808 @@
+        # they are valid before erroring out at the _convert_type stage
+        values = [
+            (self.schema, self.optional_top_level_keys),
+            (self.envelope_schema, self.optional_envelope_keys)
+        ]
+
+        return all(optionals.issubset(schema) for schema, optionals in values)
+
+    def _add_parse_result(self, record, valid, envelope):
+        """Add the result of parsing to the proper array, updating the envelope in the process
+
+        Args:
+            record (dict): Value of the parsed (or unparsed, if failed) record
+            valid (bool): True if the parsing was successful, False otherwise
+            envelope (dict): Extracted envelope to be added to the parsed record
+        """
+        if not valid:
+            self._invalid_parses.append(record)
+            return
+
+        self._apply_envelope(record, envelope)
+        self._add_optional_keys(record, self.schema, self.optional_top_level_keys)
+
+        self._valid_parses.append(record)
+
+    def _extract_envelope(self, payload):
+        """Extract envelope key/values from the original payload
+
+        Args:
+            payload (dict): The original record to pull envelope keys from
+
+        Returns:
+            dict: Key/values extracted from the log to be used as the envelope
+        """
+        if not self.envelope_schema:
+            return
+
+        LOGGER.debug('Extracting envelope keys')
+        return {
+            key: payload[key]
+            for key in self.envelope_schema
+            if key in payload  # This is fine since some of these may be optional
+        }
+
+    def _json_path_records(self, payload):
+        """Extract records from the original json payload using a provided JSON path
+
+        Args:
+            payload (dict): The parsed json data
+
+        Returns:
+            list: A list of JSON records extracted via JSON path
+        """
+        # Handle jsonpath extraction of records
+        LOGGER.debug('Parsing records with JSONPath: %s', self.json_path)
+
+        return jmespath.search(self.json_path, payload)
+
+    def parse(self, data):
+        """Main parser method to be handle parsing of the passed data
+
+        Args:
+            data (str|dict): Data to be parsed.
+
+        Returns:
+            bool: True if any records were parsed using this schema, False otherwise
+        """
+        # Ensure the schema is defined properly. Invalid schemas will not be used
+        if not self._validate_schema():
+            LOGGER.error(
+                'Schema definition is not valid (%s):\n%s', self._schema_type, self.schema)
+            return False
+
+        data_copy = None
+        # If the data is a mutable object, copy it since
+        # the parse(s) can highly mutate the input
+        if isinstance(data, dict):
+            data_copy = deepcopy(data)
+
+        # If the data is a string, attempt to load it to json,
```

Comment:
why do we do this in the base parser?

Reply:
This is done here because there are options for parsing csv data embedded in json. front loading the json loading here ensures that the CSV parser get's the expected data